### ISSUE TITLE
Create a "core" artifact

### ISSUE THREAD

**Issue by jimschubert** [OpenAPITools/openapi-generator#845]
##### Description

There's currently no clearly defined "interface" between templates, extension points, or code generation.

If we extract model definitions and interfaces to a "core" artifact, this would reduce noise for those who want to extend and/or contribute to the project.

This would ideally be done after other work in the [Separation of Concerns](https://github.com/OpenAPITools/openapi-generator/projects/5) project in this repo.

##### Related PRs

* #503
* #837
* #840 
* #841 
* #842 (maybe) 
* #843 
* #844 

##### Suggest a fix/enhancement

From #503:

![image](https://user-images.githubusercontent.com/109659/44302995-2ec9b300-a303-11e8-87a6-c9f7238ab009.png)

In the above diagram, we'd include any models (codegen, options, template) or interfaces which may be consumed by users for creating custom generators or plugins. **NOTE** Template Definition above refers to a strongly-typed model that would replace the current `Map<String, Object>` (not the resources defining templates); work for this is already started in #837.

**Comment by jmini** [comment]
I am not sure how I can add an idea to the list, but I think we should remove using `System.setProperty(..)` from our code.

It is ok to read them (requested for backward compatibility), but setting the values is a bad practice.
At the end they are some inputs options for the `DefaultGenerator`

Related issue: I have described how System properties are set by `CodegenConfigurator.setSystemProperties()` here: https://github.com/OpenAPITools/openapi-generator/issues/811#issuecomment-415313725

Should I open a new issue for this task?

**Comment by jimschubert** [comment]
@jmini yeah I think that should be another card. Work from this card will be a breaking change, but removing setProperty should be done as soon as possible, imho.

I would append that setProperty is used in the Gradle plugin, I think it's fine in this case. It shouldn't exist in the generator packages, though, because these can be consumed by other artifacts or tools.

**Comment by jmini** [comment]
Removal of `System.setProperty(..)` is now tracked in https://github.com/OpenAPITools/openapi-generator/issues/1597

**Comment by jimschubert** [comment]
Template Engine component is done via #6357

**Comment by jimschubert** [comment]
"Global Options" and "Generation Options" components are completed via #2946. I'd updated these terms to "Workflow Settings" (rather than Global Options) and "Generator Settings" rather than "Generation Options" to better differentiate between the touch points.
